### PR TITLE
Implemented MosaicRestrictionTransactionService

### PR DESCRIPTION
**Pull request by fboucquez**
New mosaic restriction service supports aliases
Implemented AliasService
Improved MetadataTransactionService so unresolved mosaic ids are supported (alias support)

Chat about the implemenation

https://nem2.slack.com/archives/CEZKUE4KB/p1576101956352300

### CODE REVIEW

File: integration-tests/src/test/java/io/nem/sdk/infrastructure/MosaicGlobalRestrictionIntegrationTest.java [hunk 1]
```diff
@@ -40,19 +35,22 @@
...
     @EnumSource(RepositoryType.class)
     void createMosaicGlobalRestrictionAndValidateEndpoints(RepositoryType type)
         throws InterruptedException {
 
         //1) Create a new mosaic
-        MosaicId mosaicId = createMosaic(type);
-        NamespaceId mosaicAlias = setMosaicAlias(type, mosaicId,
-            ("mosaicAlias" + mosaicId.getIdAsHex()).toLowerCase());
+
+        String masaicAliasName = "MosaicRestrictionServiceIntegrationTest_createMosaicGlobalRestrictionAndValidateEndpoints"
```
Comment (dgarcia360) [inline integration-tests/src/test/java/io/nem/sdk/infrastructure/MosaicGlobalRestrictionIntegrationTest.java]: mosaic*AliasName

File: sdk-core/src/main/java/io/nem/sdk/infrastructure/AliasServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                });
+        } else {
+            return Observable.just((MosaicId) unresolvedMosaicId);
+        }
+
+    }
+
+    @Override
+    public Observable<Address> resolveAddress(UnresolvedAddress unresolvedAddress) {
+        if (unresolvedAddress.isAlias()) {
```
Comment (fboucquez) [inline sdk-core/src/main/java/io/nem/sdk/infrastructure/AliasServiceImpl.java]: @dgarcia360, are the mosaic Id and address resolution? These resolutions are intended for new transactions, for all transactions we have the recipient resolution implemented in the TransactionService.

**Comment by fboucquez** [comment]
Fixed the deadline epoch offset here

**Comment by fboucquez** [comment]
@evias @dgarcia360 @rg911 

The last question is if we want the services to support aliases by resolving the aliases themselves.

What do you think? If not, I'll remove the alias service from the metadata/mosaic service and just allow real mosaic id and addresses.

File: integration-tests/src/test/java/io/nem/sdk/infrastructure/MosaicRestrictionServiceIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+    }
+
+
+    @ParameterizedTest
+    @EnumSource(RepositoryType.class)
+    void createUpdateMosaicAddressRestrictionTransactionFactoryUsingAlias(RepositoryType type)
+        throws InterruptedException {
+
+        //1) Create mosaic and alias
+        String masaicAliasName = "createUpdateMosaicAddressRestrictionTransactionFactoryUsingAlias"
```
Comment (dgarcia360) [inline integration-tests/src/test/java/io/nem/sdk/infrastructure/MosaicRestrictionServiceIntegrationTest.java]: mosaic*

File: sdk-core/src/main/java/io/nem/sdk/infrastructure/AliasServiceImpl.java (hunk 2 above)
Comment (dgarcia360) [reply sdk-core/src/main/java/io/nem/sdk/infrastructure/AliasServiceImpl.java]: It looks correct

File: sdk-core/src/main/java/io/nem/sdk/infrastructure/MosaicRestrictionTransactionServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+                })));
+    }
+
+    @Override
+    public Observable<MosaicAddressRestrictionTransactionFactory> createMosaicAddressRestrictionTransactionFactory(
+        UnresolvedMosaicId unresolvedMosaicId, BigInteger restrictionKey,
+        UnresolvedAddress unresolvedTargetAddress, BigInteger restrictionValue) {
+
+        return Observable
+            .combineLatest(networkTypeObservable, aliasService.resolveMosaicId(unresolvedMosaicId),
```
Comment (dgarcia360) [inline sdk-core/src/main/java/io/nem/sdk/infrastructure/MosaicRestrictionTransactionServiceImpl.java]: Why the mosaic is being resolved first? I believe you could pass the UnresolvedMosaic (either MosaicId or NamespaceId) directly to the MosaicAddressRestrictionTransaction.

File: sdk-core/src/main/java/io/nem/sdk/infrastructure/MosaicRestrictionTransactionServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                aliasService.resolveAddress(unresolvedTargetAddress),
```
Comment (dgarcia360) [inline sdk-core/src/main/java/io/nem/sdk/infrastructure/MosaicRestrictionTransactionServiceImpl.java]: ^Same with address resolution

File: sdk-core/src/main/java/io/nem/sdk/infrastructure/MosaicRestrictionTransactionServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        this.networkTypeObservable = repositoryFactory.getNetworkType();
+        this.aliasService = new AliasServiceImpl(repositoryFactory);
+    }
+
+
+    @Override
+    public Observable<MosaicGlobalRestrictionTransactionFactory> createMosaicGlobalRestrictionTransactionFactory(
+        UnresolvedMosaicId unresolvedMosaicId, BigInteger restrictionKey,
+        BigInteger restrictionValue, MosaicRestrictionType restrictionType) {
+        return aliasService.resolveMosaicId(unresolvedMosaicId).flatMap(mosaicId ->
```
[the remark this answers is not in the document]
Comment (dgarcia360) [reply sdk-core/src/main/java/io/nem/sdk/infrastructure/MosaicRestrictionTransactionServiceImpl.java]: Solved, " rest needs to call the rest endpoint to know the previous restriction/metadata value
and the rest endpoints only accept real address/mosaic id"

[the remark this answers is not in the document]
Comment (fboucquez) [reply sdk-core/src/main/java/io/nem/sdk/infrastructure/MosaicRestrictionTransactionServiceImpl.java]: SDK needs to call the rest endpoints to know the previous restriction/metadata value. Rest's endpoints only accept real address/mosaic id. Once the previous value is known, the transaction can be created using the provided alias.

File: integration-tests/src/test/java/io/nem/sdk/infrastructure/MosaicRestrictionServiceIntegrationTest.java (hunk 3 above)
Comment (fboucquez) [reply integration-tests/src/test/java/io/nem/sdk/infrastructure/MosaicRestrictionServiceIntegrationTest.java]: Fixed

File: sdk-core/src/main/java/io/nem/sdk/infrastructure/MosaicRestrictionTransactionServiceImpl.java (hunk 4 above)
Comment (fboucquez) [reply sdk-core/src/main/java/io/nem/sdk/infrastructure/MosaicRestrictionTransactionServiceImpl.java]: SDK needs to call the rest endpoints to know the previous restriction/metadata value. Rest's endpoints only accept real address/mosaic id. Once the previous value is known, the transaction can be created using the provided alias.

File: sdk-core/src/main/java/io/nem/sdk/infrastructure/MosaicRestrictionTransactionServiceImpl.java (hunk 5 above)
Comment (fboucquez) [reply sdk-core/src/main/java/io/nem/sdk/infrastructure/MosaicRestrictionTransactionServiceImpl.java]: Same problem. The endpoint to get the current Mosaic address restriction only accepts real addresses.

https://nemtech.github.io/nem2-openapi/#operation/getMosaicAddressRestriction

**Comment by fboucquez** [comment]
We have decided to go with alias resolution in the service when and only when necessary (when the parameters send are aliases/namespace ids)

File: integration-tests/src/test/java/io/nem/sdk/infrastructure/AccountMetadataIntegrationTest.java [hunk 7]
```diff
@@ -64,7 +64,7 @@ public void addMetadataToAccount(RepositoryType type) throws InterruptedExceptio
         Assertions.assertEquals(transaction.getScopedMetadataKey(),
             processedTransaction.getScopedMetadataKey());
 
-        sleep(2000);
+       // sleep(2000);
```
Comment (rg911) [inline integration-tests/src/test/java/io/nem/sdk/infrastructure/AccountMetadataIntegrationTest.java]: remove the comments?

Comment (fboucquez) [reply integration-tests/src/test/java/io/nem/sdk/infrastructure/AccountMetadataIntegrationTest.java]: Fixed

**Comment by fboucquez** [comment]
@dgarcia360 @evias @rg911 is it possible to approve this? It seems like everybody is happy...

**Pull request merged**